### PR TITLE
Feature/mac support

### PR DESCRIPTION
**Pull request by ZvonimirMedak**
## Summary
Added macOS support with a new SwiftUI example

## Changes
- Podspec updated (removed CustomLocation from default podspec, added macOS support)
- Added a new SwiftUI example (we will most likely need to add the SPM every time the project is downloaded since currently it has my local path)
- Added macOS target to the UIKit example
- Added macOS support to the DeviceTool
- exposed SentinelTabBarView to be public
- refactored Preferences to use PreferenceSection so that the users can't add other items than Toggle ones (We will introduce other ones later on or maybe we can set it to toggle/custom)

### Type

- [X] **Feature**: This pull request introduces a new feature.
- [ ] **Bug fix**: This pull request fixes a bug.
- [X] **Refactor**: This pull request refactors existing code.
- [X] **Documentation**: This pull request updates documentation.
- [ ] **Other**: This pull request makes other changes.

#### Additional information

- [X] This pull request introduces a **breaking change**.

### Description

I've mentioned everything in the changes section

## Checklist

- [X] I have performed a self-review of my own code.
- [X] I have tested my changes, including edge cases.
- [X] I have added necessary tests for the changes introduced (if applicable).
- [X] I have updated the documentation to reflect my changes (if applicable).

## Additional notes

After this gets reviewed, I would like to try to push team members to try out the new version where possible so we can test it out. Also, I would like to improve the current state of the navigation on macOS

### CODE REVIEW

File: Example/MacOSExample/AppDelegate.swift
Comment (nikolamajcen) [inline Example/MacOSExample/AppDelegate.swift:1]: I would recommend to use buildable folders and re-arrange file structure a bit. I'm okay with you to continue using e.g. `Example for iOS` and `Example for macOS` folder naming like it was already used (but they were groups).

File: Example/MacOSExample/AppDelegate.swift [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    var window: NSWindow?
+
+    func applicationDidFinishLaunching(_ aNotification: Notification) {
+        // Insert code here to initialize your application
+        setupSentinel()
+    }
+
+    func applicationWillTerminate(_ aNotification: Notification) {
+        // Insert code here to tear down your application
+    }
```
Comment (nikolamajcen) [inline Example/MacOSExample/AppDelegate.swift:35]: Remove methods that are not used.

File: Example/MacOSExample/AppDelegate.swift [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    func applicationSupportsSecureRestorableState(_ app: NSApplication) -> Bool {
+        return true
+    }
+
```
Comment (nikolamajcen) [inline Example/MacOSExample/AppDelegate.swift:40]: Remove redundant new line.

File: Example/MacOSExample/AppDelegate.swift
Comment (nikolamajcen) [reply Example/MacOSExample/AppDelegate.swift:1]: Additionally, try to use the same naming definition for targets; e.g. use `Example-iOS` and `Example-macOS`.

File: Example/MacOSExample/ViewController.swift [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        super.viewDidLoad()
+
+        // Do any additional setup after loading the view.
+    }
+
+    override var representedObject: Any? {
+        didSet {
+        // Update the view, if already loaded.
+        }
+    }
```
Comment (nikolamajcen) [inline Example/MacOSExample/ViewController.swift:25]: Remove methods and properties which overrides are empty.

File: Example/Podfile [hunk 4]
```diff
@@ -10,3 +11,9 @@ target 'Sentinel_Example' do
     inherit! :search_paths
   end
 end
+
+target 'MacOSExample' do
```
Comment (nikolamajcen) [inline Example/Podfile:15]: Update target names for both iOS and macOS - as mentioned in the previous comment.

File: Example/Sentinel/AppDelegate.swift [hunk 5]
```diff
@@ -67,31 +66,25 @@ private extension AppDelegate {
         )
     }
 
-    var optionSwitchItems: [ToolTableSection] {
+    var optionSwitchItems: [PreferencesTool.Section] {
         [
             .init(
```
Comment (nikolamajcen) [inline Example/Sentinel/AppDelegate.swift:71]: For better readability use class names instead of `.init`. Also, update that through the whole codebase if applicable.

We are trying to make examples and our code as readable as possible.

File: Package.swift [hunk 6]
```diff
@@ -3,9 +3,19 @@
...
 import PackageDescription
 
+#if os(macOS)
+let exclusions: [String] = [
+    "Classes/CustomLocation",
+    "Classes/EmailSender"
+]
+#else
+let exclusions: [String] = []
+#endif
```
Comment (nikolamajcen) [inline Package.swift:13]: Nice 👍

Still, try to define the variable like `excludedSources` to make it clearer (we are not excluding resources here, e.g. images).

File: Sentinel.podspec [hunk 7]
```diff
@@ -25,6 +25,7 @@ Pod::Spec.new do |s|
   s.platform = :ios
   s.swift_version = '5.0'
   s.ios.deployment_target = '14.0'
+  s.osx.deployment_target = '12.0'
```
Comment (nikolamajcen) [inline Sentinel.podspec:28]: Not sure if this was added, but please update the `README.md` with that information.

File: Sentinel/Classes/Core/Internal/DeviceTool.swift [hunk 8]
```diff
@@ -52,11 +47,38 @@ extension DeviceTool {
...
+            .init(title: "UUID", value: hardwareDeviceUUID ?? "???")
+        ])
+        #else
+        CustomInfoTool.Section(title: "Device", items: [
+            .init(title: "Model", value: UIDevice.current.model),
+            .init(title: "Name", value: UIDevice.current.name),
+            .init(title: "System version", value: systemVersion),
+            .init(title: "UUID", value: UIDevice.current.identifierForVendor?.uuidString ?? "???"),
+            .init(title: "Battery state", value: batteryState),
+            .init(title: "Proximity state", value: UIDevice.current.proximityState ? "Close" : "Far")
```
Comment (nikolamajcen) [inline Sentinel/Classes/Core/Internal/DeviceTool.swift:65]: Replace `.init` methods with real class names.

Also, additionally, if the objects of the array are the same, you could instead return the array from a method instead creating the whole object inside the `if-else` branches. Even better approach here would be a factory method or abstract factory (due to other methods), but I believe this can be added and improved in other PRs.

File: Sentinel/Classes/Core/Trigger.swift [hunk 9]
```diff
@@ -18,16 +21,19 @@ public protocol Trigger {
...
-    public static var shake: Trigger { ShakeTrigger() }
 
-    /// The trigger type which is triggered on the screenshot event.
-    public static var screenshot: Trigger { ScreenshotTrigger() }
 
     /// The trigger type which is triggered on the specified notification name.
     public static func notification(forName name: Notification.Name) -> Trigger {
         NotificationTrigger(notificationName: name)
     }
+    #if !os(macOS)
```
Comment (nikolamajcen) [inline Sentinel/Classes/Core/Trigger.swift:30]: Would it be better to use `#if os(iOS)`? Because with this, e.g. `tvOS` or `watchOS` in the future would support this. I know we will not add the possibility for that, but it is easier to work with constrained things than the opposite.

Comment (nikolamajcen) [reply Sentinel/Classes/Core/Trigger.swift:30]: Also, add a note to the comment (or discussion), on which platforms are those methods available.

File: Sentinel/Classes/TextEditing/TextEditingToolView.swift [hunk 10]
```diff
@@ -19,7 +19,9 @@ struct TextEditingToolView: View {
 
             Button(action: {
                 viewModel.didPressSave(viewModel.value)
-                presentationMode.wrappedValue.dismiss()
+                #if !os(macOS)
```
Comment (nikolamajcen) [inline Sentinel/Classes/TextEditing/TextEditingToolView.swift:22]: Can we use `os(iOS)` here as well?

File: Sentinel/Classes/UserDefaults/UserDefaultsToolView.swift [hunk 11]
```diff
@@ -17,15 +17,22 @@ struct UserDefaultsToolView: View {
...
                         UIPasteboard.general.string = viewModel.value
+                        #endif
                     })
                 }))
                 .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
 
             Button(action: {
                 viewModel.didPressDelete()
-                presentationMode.wrappedValue.dismiss()
+                #if !os(macOS)
```
Comment (nikolamajcen) [inline Sentinel/Classes/UserDefaults/UserDefaultsToolView.swift:31]: Same here as well.

File: SwiftUIExample/SwiftUIExample/ContentView.swift
Comment (nikolamajcen) [inline SwiftUIExample/SwiftUIExample/ContentView.swift:1]: Regarding the SwiftUI example, I think this is a bit redundant in the current form. We have a new project that we need to maintain, while the implementation is different only for view part (app delegate counterpart is same).

I would rather propose to add the SwiftUI code in the existing example in one of the provided forms:

* Create two screens (for iOS and macOS targets) - one with UIKit and another with SwiftUI examples (with summoning Sentinel)
* Create a new target that will work on both iOS and macOS

**Review by nikolamajcen — CHANGES_REQUESTED** [review]
This is such a great work 👍 I'm glad that I can finally run Sentinel on macOS.

Based on your changes, I've added some comments so please check them out.

**Comment by nikolamajcen** [comment]
Also, one additional note; I've seen you've done a lot of work, but you've mixed that into bigger PRs, e.g. podspec update, macOS support, preferences update. On the other PR there are multiple topics as well.

Try to separate them in future PR for easier review process and separation of concerns. The work you did is great, but this can make things easier to understand and will produce quicker review times.

File: Sentinel.podspec (hunk 7 above)
Comment (ZvonimirMedak) [reply Sentinel.podspec:28]: I would go through the readme again once we wrap up everything since there were quite a lot of changes, and I wouldn't want to do it multiple times if something changes before we release

File: SwiftUIExample/SwiftUIExample/ContentView.swift
Comment (ZvonimirMedak) [reply SwiftUIExample/SwiftUIExample/ContentView.swift:1]: I have added a new SwiftUI screen to both MacOS and iOS example target

**Review by nikolamajcen — APPROVED** [review]
Hey @ZvonimirMedak , nice work, for real! This is amazing.

Just one change and I think we are good to go; in the root folder I can se SwiftUIExample folder which I think can be removed in this scenario. Once you get rid of that, you can merge - I'm approving it already.
